src/utilities/alfalfa/AlfalfaComponent.hpp

```diff
@@ -0,0 +1,59 @@
+#ifndef ALFALFA_COMPONENT_HPP
+#define ALFALFA_COMPONENT_HPP
+
+#include <json/json.h>
+
+#include "../UtilitiesAPI.hpp"
+
+namespace openstudio {
+namespace alfalfa {
+  enum Capability
+  {
+    Input = 1,
+    Output = 2
+  };
+
+  inline Capability operator|(const Capability& a, const Capability& b) {
+    return static_cast<Capability>(static_cast<int>(a) | static_cast<int>(b));
+  }
+
+  inline Capability operator&(const Capability& a, const Capability& b) {
+    return static_cast<Capability>(static_cast<int>(a) & static_cast<int>(b));
+  }
+
+  class UTILITIES_API AlfalfaComponent
+  {
+   public:
+    AlfalfaComponent(const std::string& type, const Capability capabilities);
+
+    /**
+     * Get if the component can be used as an Input to the model
+     */
+    bool canInput() const;
+
+    /**
+     * Get if the component can be used as an Output from the model
+     */
+    bool canOutput() const;
+
+    bool operator==(const AlfalfaComponent& rhs) const;
+
+    /**
+     * A string which reflects the type of the component
+     */
+    std::string type;
+
+    /**
+     * The raw input/output capabilities of the component
+     */
+    Capability capabilities;
+
+    /**
+     * JSON encoded parameters of the component
+     */
+    Json::Value parameters;
```

Comment:
yeah, this is the part I'm least comfortable with. I want the fields of the of each of the independent children of `AlfalfaComponent` to be accessible from an instance of the superclass as there are cases where a method may just return an `AlfalfaComponent` type when it can be any of the children. It seems like the pattern used in OS is the pImpl, but I though that was a bit complex for something so simple.